pandas/core/dtypes/cast.py

```diff
@@ -1307,9 +1309,16 @@ def maybe_infer_to_datetimelike(
        leave inferred dtype 'date' alone
 
     """
-    # TODO: why not timedelta?
     if isinstance(
-        value, (ABCDatetimeIndex, ABCPeriodIndex, ABCDatetimeArray, ABCPeriodArray)
+        value,
+        (
+            ABCDatetimeArray,
+            ABCDatetimeIndex,
```

Comment:
@jbrockmendel WDYT?

Reply:
I think this can be simplified to something like:

```
if hasattr(value, "dtype") and not is_object_dtype(value.dtype):
    return value
```

(might need to do an `extract_array` in there).

Also seems likely we can narrow down what gets passed in to this function

Reply:
> I think this can be simplified to something like:
> 
> ```
> if hasattr(value, "dtype") and not is_object_dtype(value.dtype):
>     return value
> > ```
Pushed this now and will post asvs

Reply:
> Also seems likely we can narrow down what gets passed in to this function

Current use cases:
```
pandas/core/arrays/categorical.py:322:            values = maybe_infer_to_datetimelike(values, convert_dates=True)
pandas/core/frame.py:3829:                value = maybe_infer_to_datetimelike(value)
pandas/core/dtypes/cast.py:1529:            value = maybe_infer_to_datetimelike(value)
pandas/core/internals/construction.py:218:            dvals_list = [maybe_infer_to_datetimelike(row) for row in values]
pandas/core/internals/construction.py:231:            datelike_vals = maybe_infer_to_datetimelike(values)
```